flax/linen/module.py

```diff
@@ -432,11 +451,15 @@ def __setattr__(self, name: str, val: Any):
     # Submodules are being defined and attached in setup()
     else:
       val = _freeze_attr(val)
+      already_processed_modules = set()
       for suffix, subvalue in _get_suffix_value_pairs(val):
         if isinstance(subvalue, Module):
           if not self._state.in_setup:
             raise ValueError(
                 "You can only assign submodules to self in setup().")
+          # If a modules appears multiple times in the pytree, attach it only once.
+          if id(subvalue) in already_processed_modules:
```

Comment:
So is it intended that:
```
a = Dense()
self.a = a
self.b = b # raises error
```
but
```
a = Dense()
self.a = [a, a]
```
does not raise an error?

Reply:
Sorry, I should have commented on this change, or even split it out for later -- it's actually completely orthogonal to the rest - I just noticed that the latter example you give does in fact fail with an exception now.  I could imagine someone making a list w. replicated layers in setup() to do something like e.g. "universal transformer" which is a stack of shared param layers.

In first example, I'm assuming you meant
```
a = Dense()
self.a = a
self.b = a # raises error
```
That is another interesting example that would still error-out in this case.

Now I'm thinking that perhaps we should stick to the current "unique-names on self in setup()" policy and only try to "fix" this if there's an actual problem that users encounter.  At the very least I think I'll split this out to discuss later/elsewhere, as it's completely separate from the other changes.

Reply:
I've removed this from this PR - it's a complicated issue worth covering later.